src/java/org/apache/cassandra/index/SecondaryIndexManager.java

```diff
@@ -132,9 +144,20 @@ public class SecondaryIndexManager implements IndexRegistry, INotificationConsum
 {
     private static final Logger logger = LoggerFactory.getLogger(SecondaryIndexManager.class);
 
+    // executes index status propagation task asynchronously to avoid potential deadlock on SIM
+    private static final ExecutorPlus statusPropagationExecutor = executorFactory().withJmxInternal()
+                                                                                   .sequential("StatusPropagationExecutor");
+    // used to produce a status string from the current endpoint states in JSON format for Gossip propagation
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
     // default page size (in rows) when rebuilding the index for a whole partition
     public static final int DEFAULT_PAGE_SIZE = 10000;
 
+    /**
+     * A map of per-endpoint index statuses: the key of inner map is the identifier "keyspace.index"
+     */
+    public static final Map<InetAddressAndPort, Map<String, Index.Status>> peerIndexStatus = new ConcurrentHashMap<>();
```

Comment:
I like this idea a lot. It removes a block of code from `SecondaryIndexManager` that didn't strictly belong there. I have gone with the name `IndexStatusManager`. I have moved `filterForQuery` into this class because it is more related to index statuses than index management. 

I have gone with `peerIndexStatus` being a `HashMap` with access to it `synchronized`. It's a difficult call on whether this is correct or using a `ConcurrentHashMap` and having the methods non-synched is better. I have gone with the `synchronized` methods to avoid and potential race conditions in the methods themselves. The `peerIndexStatus` map itself does not need to be concurrent because it is only now accessed from within the `synchronized` methods.